### ISSUE TITLE
Disable type assertion of errors when used with assignment in a switch statement

### ISSUE THREAD

**Issue by avinash-fam** [polyfloyd/go-errorlint#43]
My code looks something like this

```
type FooError struct {
	Code       string
	Err        string
	HttpStatus int
}

type BarError struct {
	Code       string
	Err        []string
	HttpStatus int
}

var ErrUnexpected = FooError{
	Code:       "E001",
	Err:        "Unexpected error.",
	HttpStatus: http.StatusInternalServerError,
}

func SomeFunction(){
	switch errPayload := err.(type) {    // error lint here
	case customerror.FooError:
		SendJSONResponse(w, errPayload.HttpStatus, errPayload)
	case customerror.BarError:
		SendJSONResponse(w, errPayload.HttpStatus, errPayload)
	default:
		errGus := customerror.ErrUnexpected
		SendJSONResponse(w, errGus.HttpStatus, errGus, logger)
	}
}
```

I get the following linting error for the marked line `type switch on error will fail on wrapped errors. Use errors.As to check for specific errors`.

I don't think we can use `errors.As()` in switch statement to iterate over cases since the second argument of `As()` requires a target of known type. And I don't know the type at the highlighted line.

**Comment by polyfloyd** [comment]
Hi! I am not sure I follow your question. You do know the type at the switch statement, they are in the individual `case`s.

You could do something like:
```
var foo customerror.FooError
var bar customerror.BarError
if errors.As(err, &foo) {
	SendJSONResponse(w, errPayload.HttpStatus, errPayload)
} else if errors.As(err, &bar) {
	SendJSONResponse(w, errPayload.HttpStatus, errPayload)
} else {
	errGus := customerror.ErrUnexpected
	SendJSONResponse(w, errGus.HttpStatus, errGus, logger)
}
```

**Comment by avinash-fam** [comment]
Thanks for the reply, but won't the above stuff give another linting error: `ifElseChain: rewrite if-else to switch statement`?

The whole point of this issue is to do one of the following:
1. do not throw lint error at the line marked in my original comment
2. do not throw lint error at the `if errors.As(err, &foo) {` line in your snippet

**Comment by polyfloyd** [comment]
This does not look like an error from a linter that I have enabled myself.

It may want you to rewrite it to a switch with bool expression as cases. What about below then?
```go
var foo customerror.fooerror
var bar customerror.barerror
switch {
case errors.as(err, &foo):
	sendjsonresponse(w, errpayload.httpstatus, errpayload)
case errors.as(err, &bar):
	sendjsonresponse(w, errpayload.httpstatus, errpayload)
default:
	errgus := customerror.errunexpected
	sendjsonresponse(w, errgus.httpstatus, errgus, logger)
}
```

**Comment by kolyshkin** [comment]
@avinash-fam the warning that the linter is giving is legit -- your code won't work on wrapped errors. If you're sure what you're doing (and you absolutely do not expect a wrapped error in this code), suppress the linter warning (for example, if you use golangci-lint, with something like this):

```go
	switch errPayload := err.(type) { //nolint:errorlint // err is never wrapped.
```

If you are not sure, switch to using `errors.As` as suggested above.
